numpy/fft/fftpack.py

```diff
@@ -86,7 +92,8 @@ def _raw_fft(a, n=None, axis=-1, init_function=fftpack.cffti,
     # As soon as we put wsave back into the cache, another thread could pick it
     # up and start using it, so we must not do this until after we're
     # completely done using it ourselves.
-    fft_cache[n].append(wsave)
+    with _cache_lock:
+        fft_cache[n].append(wsave)
```

Comment:
I think you could get rid of `setdefault` and just have  `getitem` and `setitem` functions. The internal logic would be much more apparent that way.